node/src/components/consensus/validator_change.rs

```diff
@@ -24,27 +26,219 @@ impl ValidatorChange {
     pub(super) fn era_changes<I>(
         era0: &Era<I>,
         era1: &Era<I>,
+    ) -> Vec<(PublicKey, ValidatorChange)> {
+        let era0_metadata = EraMetadata::from(era0);
+        let era1_metadata = EraMetadata::from(era1);
+        Self::validator_changes(era0_metadata, era1_metadata)
+    }
+
+    fn validator_changes(
+        era0_metadata: EraMetadata,
+        era1_metadata: EraMetadata,
     ) -> Vec<(PublicKey, ValidatorChange)> {
         let mut changes = Vec::new();
-        for pub_key in era0.validators().keys() {
-            if !era1.validators().contains_key(pub_key) {
+        for pub_key in era0_metadata.validators.keys() {
+            if !era1_metadata.validators.contains_key(pub_key) {
                 changes.push((pub_key.clone(), ValidatorChange::Removed));
             }
         }
-        for pub_key in era1.consensus.validators_with_evidence() {
+        for pub_key in era1_metadata.seen_as_faulty {
             changes.push((pub_key.clone(), ValidatorChange::SeenAsFaulty));
         }
-        for pub_key in era1.validators().keys() {
-            if !era0.validators().contains_key(pub_key) {
+        for pub_key in era1_metadata.validators.keys() {
+            if !era0_metadata.validators.contains_key(pub_key) {
                 changes.push((pub_key.clone(), ValidatorChange::Added));
             }
-            if era1.faulty.contains(pub_key) && !era0.faulty.contains(pub_key) {
+            if era1_metadata.faulty.contains(pub_key) && !era0_metadata.faulty.contains(pub_key) {
                 changes.push((pub_key.clone(), ValidatorChange::Banned));
             }
-            if era1.cannot_propose.contains(pub_key) && !era0.cannot_propose.contains(pub_key) {
+            if era1_metadata.cannot_propose.contains(pub_key)
+                && !era0_metadata.cannot_propose.contains(pub_key)
+            {
                 changes.push((pub_key.clone(), ValidatorChange::CannotPropose));
             }
         }
         changes
     }
 }
+
+#[derive(Clone)]
+struct EraMetadata<'a> {
+    validators: &'a BTreeMap<PublicKey, U512>,
+    seen_as_faulty: Vec<PublicKey>,
```

Comment:
I think we could store references to `PublicKey` here, like so:
```
seen_as_faulty: Vec<&'a PublicKey>,
```

hence, avoid cloning in `from()`, like so:
```
        let seen_as_faulty = era
            .consensus
            .validators_with_evidence()
            .into_iter()
            .collect();
```

Please consider changing the code, even though I'm not sure what savings we are fighting for here :)

Reply:
I see this is already fixed by @Fraser999 in https://github.com/darthsiroftardis/casper-node/pull/12 👍